### PR TITLE
improve android emulator detection

### LINKED ISSUE #16165: [expo-device] isDevice returns true when app is running in emulator
### Summary

What I'm expecting is to detect what kind of device the app is running on and then open/close notification service based on this. In simulator, expo-device detects correctly and isDevice returns -> false, but in an android emulator it's true. It should be false as well.

I'm running project Mac M1 chip, this can be another reason why I see this. 

Here is another link about this problem but from the other way around.
[https://forums.expo.dev/t/expo-device-package-api-bug/56802](url)

Screenshot 2022-02-02 at 14 41 27

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

bare

### What platform(s) does this occur on?

Android

### SDK Version (managed workflow only)

_No response_

### Environment

Expo CLI 5.0.3 environment info:
    System:
      OS: macOS 12.1
      Shell: 5.8 - /bin/zsh
    Binaries:
      Node: 14.17.3 - /usr/local/bin/node
      Yarn: 1.22.17 - /usr/local/bin/yarn
      npm: 6.14.13 - /usr/local/bin/npm
    Managers:
      CocoaPods: 1.11.2 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 21.2, iOS 15.2, macOS 12.1, tvOS 15.2, watchOS 8.3
      Android SDK:
        API Levels: 29, 30, 31, 32
        Build Tools: 29.0.2, 30.0.2, 30.0.3, 32.0.0
        System Images: android-31 | Google APIs ARM 64 v8a, android-31 | Google Play Intel x86 Atom_64, android-32 | Google APIs ARM 64 v8a
        Android NDK: 22.1.7171670
    IDEs:
      Android Studio: 2020.3 AI-203.7717.56.2031.7935034
      Xcode: 13.2.1/13C100 - /usr/bin/xcodebuild
    npmPackages:
      expo: ~43.0.2 => 43.0.4 
      react: 17.0.1 => 17.0.1 
      react-dom: 17.0.1 => 17.0.1 
      react-native: 0.64.3 => 0.64.3 
      react-native-web: 0.17.1 => 0.17.1 
    npmGlobalPackages:
      expo-cli: 5.0.3
    Expo Workflow: bare

### Reproducible demo

It works with snacks as normal.

### PR DESCRIPTION
**Pull request by kbrandwijk**
# Why

Existing Android Emulator detection is very basic, and leads to false positives/negatives in detection. This PR aims to improve that detection across the board.

# How

The new logic is adapted from react-native-device-info (https://github.com/react-native-device-info/react-native-device-info/blob/master/android/src/main/java/com/learnium/RNDeviceInfo/RNDeviceModule.java#L225). Changes are committed per package, so this can also be cherry picked if needed.

# Test Plan

All existing tests pass, but whether this logic is covered is questionable. Might need additional manual testing, but I will need some directions, as it touches quite a few packages. 

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: android/ReactAndroid/src/main/java/com/facebook/react/modules/systeminfo/AndroidInfoHelpers.java [hunk 1]
```diff
@@ -37,7 +37,29 @@ private static boolean isRunningOnGenymotion() {
     }
 
     private static boolean isRunningOnStockEmulator() {
```
Comment (ide) [inline android/ReactAndroid/src/main/java/com/facebook/react/modules/systeminfo/AndroidInfoHelpers.java:39]: This code is synced in and not authored by us. Manual changes to this file will be erased.

File: packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/launcher/DevLauncherActivity.kt [hunk 2]
```diff
@@ -102,7 +102,29 @@ class DevLauncherActivity : ReactActivity(), ReactInstanceManager.ReactInstanceE
   }
 
   private val isSimulator
```
Comment (ide) [inline packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/launcher/DevLauncherActivity.kt:104]: expo-dev-launcher depends on expo-dev-menu -- one of the people who works on these modules can probably say whether expo-dev-menu could expose a Java-only method that expo-dev-launcher could use.

File: android/expoview/src/main/java/host/exp/exponent/ExpoUpdatesAppLoader.kt [hunk 3]
```diff
@@ -399,10 +399,35 @@ class ExpoUpdatesAppLoader @JvmOverloads constructor(
...
+      || Build.PRODUCT.contains("simulator")
+      || Build.BOARD.lowercase(Locale.ROOT).contains("nox")
+      || Build.BOOTLOADER.lowercase(Locale.ROOT).contains("nox")
+      || Build.HARDWARE.lowercase(Locale.ROOT).contains("nox")
+      || Build.PRODUCT.lowercase(Locale.ROOT).contains("nox")
+      || Build.SERIAL.lowercase(Locale.ROOT).contains("nox")
+      || (Build.BRAND.startsWith("generic") && Build.DEVICE.startsWith("generic"))
+
   private val clientEnvironment: String
     get() = if (Constants.isStandaloneApp()) {
```
Comment (ide) [inline android/expoview/src/main/java/host/exp/exponent/ExpoUpdatesAppLoader.kt:428]: This code is already calling into Constants -- possible to ask Constants if the app is running in an emulator?

File: packages/expo-camera/android/src/main/java/expo/modules/camera/ExpoCameraView.kt [hunk 4]
```diff
@@ -203,12 +203,37 @@ class ExpoCameraView(
 
   override fun getPreviewSizeAsArray() = intArrayOf(previewSize.width, previewSize.height)
 
+  private val isRunningOnEmulator: Boolean
```
Comment (ide) [inline packages/expo-camera/android/src/main/java/expo/modules/camera/ExpoCameraView.kt:206]: The two camera files should be able to share a function.

**Review by ide — CHANGES_REQUESTED** [review]
Several places in this code can be deduplicated in a straightforward way but ideally we'd be able to import a shared package. I'm not familiar with the best way to do that but perhaps someone else sees the answer clearly.

File: android/expoview/src/main/java/host/exp/exponent/ExpoUpdatesAppLoader.kt (hunk 3 above)
Comment (kbrandwijk) [reply android/expoview/src/main/java/host/exp/exponent/ExpoUpdatesAppLoader.kt:428]: That's the Constants class inside of expoview, not expo-constants.

File: packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/launcher/DevLauncherActivity.kt (hunk 2 above)
Comment (kbrandwijk) [reply packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/launcher/DevLauncherActivity.kt:104]: While I agree in theory, to keep the change minimal all I really did was just swap out one condition check for a slightly larger one. Your proposed refactoring goes a bit further and could potentially wait for the upcoming audit to keep this PR lean.

**Review by bbarthec — CHANGES_REQUESTED** [review]
Following @ide's idea - I'm reluctant to allow this much duplication. We can find a single point of truth that is accessible from each part of the codebase you've modified. My guess would be `expo-modules-core` as everything depends on it and it's included everywhere (not sure about dev-client/dev-menu though) 🤔 cc @tsapeta, @lukmccall

**Review by bbarthec — APPROVED** [review]
@kbrandwijk, given what you're writing I'm willing to accept the current shape of this PR, but please document somehow (by creating an Linear.app issue?) the next refactor we're expecting to happen with this `isRunningOnEmulator` check spread across the codebase.

**Comment by kbrandwijk** [comment]
> @kbrandwijk, given what you're writing I'm willing to accept the current shape of this PR, but please document somehow (by creating an Linear.app issue?) the next refactor we're expecting to happen with this `isRunningOnEmulator` check spread across the codebase.

Thanks, done! https://linear.app/expo/issue/ENG-4109/refactor-android-emulator-detection-across-packages

**Review by bbarthec — CHANGES_REQUESTED** [review]
I've looked into the CI issues, and almost every package you've modified reported similar problem: `Unresolved reference: Locale`. Have you tried compiling the project after you've applied the changes? 🤔 
Additionally, let's add missing `CHANGELOG` entries in each affected package.

**Comment by kbrandwijk** [comment]
I still have issues with my local devenv preventing me from actually building, so I've pulled this back into draft until that is resolved, and I can actually verify my own changes properly.

**Comment by keith-kurak** [comment]
Closing in favor of https://github.com/expo/expo/pull/18394

**Pull request closed without merging**